pkg/metrics/metrics.go

```diff
@@ -0,0 +1,265 @@
+	return mgr
+}
+
+// OperationStart starts a new operation
+func (opMgr *operationMetricsManager) OperationStart(key OperationKey, val OperationValue) {
+	opMgr.mu.Lock()
+	defer opMgr.mu.Unlock()
+
+	if _, exists := opMgr.cache[key]; !exists {
+		val.startTime = time.Now()
+		opMgr.cache[key] = val
+	}
+	opMgr.opInFlight.Set(float64(len(opMgr.cache)))
+}
+
+// DropOperation drops an operation
+func (opMgr *operationMetricsManager) DropOperation(op OperationKey) {
+	opMgr.mu.Lock()
+	defer opMgr.mu.Unlock()
+	delete(opMgr.cache, op)
+	opMgr.opInFlight.Set(float64(len(opMgr.cache)))
+}
+
+// RecordMetrics emits operation metrics
+func (opMgr *operationMetricsManager) RecordMetrics(opKey OperationKey, opStatus OperationStatus, driverName string) {
+	opMgr.mu.Lock()
+	defer opMgr.mu.Unlock()
+	opVal, exists := opMgr.cache[opKey]
+	if !exists {
+		// the operation has not been cached, return directly
+		return
+	}
+
+	strStatus := opStatusUnknown
+	if opStatus != nil {
+		strStatus = opStatus.String()
+	}
+
+	// if we do not know the driverName while recording metrics,
+	// refer to the cached version instead.
+	if driverName == "" || driverName == unknownDriverName {
+		driverName = opVal.Driver
+	}
+
+	operationDuration := time.Since(opVal.startTime).Seconds()
+	opMgr.opLatencyMetrics.WithLabelValues(driverName, opKey.Name, opVal.Driver, strStatus).Observe(operationDuration)
+
+	delete(opMgr.cache, opKey)
+}
+
+func (opMgr *operationMetricsManager) init() {
+	opMgr.registry = k8smetrics.NewKubeRegistry()
+	k8smetrics.RegisterProcessStartTime(opMgr.registry.Register)
+	opMgr.opLatencyMetrics = k8smetrics.NewHistogramVec(
+		&k8smetrics.HistogramOpts{
+			Subsystem: subSystem,
+			Name:      operationLatencyMetricName,
+			Help:      operationLatencyMetricHelpMsg,
+			Buckets:   metricBuckets,
+		},
+		[]string{labelDriverName, labelOperationName, labelOperationStatus},
+	)
+	opMgr.registry.MustRegister(opMgr.opLatencyMetrics)
+	opMgr.opInFlight = k8smetrics.NewGauge(
+		&k8smetrics.GaugeOpts{
+			Subsystem: subSystem,
+			Name:      operationInFlightName,
+			Help:      operationInFlightHelpMsg,
+		},
+	)
+	opMgr.registry.MustRegister(opMgr.opInFlight)
+
+	// While we always maintain the number of operations in flight
+	// for every metrics operation start/finish, if any are leaked,
+	// this scheduled routine will catch any leaked operations.
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go opMgr.scheduleOpsInFlightMetric(ctx)
+}
+
+func (opMgr *operationMetricsManager) scheduleOpsInFlightMetric(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			for range time.NewTicker(inFlightCheckInterval).C {
+				func() {
+					opMgr.mu.Lock()
+					defer opMgr.mu.Unlock()
+					opMgr.opInFlight.Set(float64(len(opMgr.cache)))
+				}()
+			}
+		}
+	}
+}
+
+func (opMgr *operationMetricsManager) PrepareMetricsPath(mux *http.ServeMux, pattern string, logger promhttp.Logger) error {
```

Comment:
This function is an external entry point and it should not deal with internal data types like `promhttp.Logger`. Instead it should either accept a `klog.Logger` or simply craft the logger adapter needed over the default `klog.Logger`.